### PR TITLE
Bazel 8 and platform flags

### PR DESCRIPTION
**Pull request by RyanDraves**
So I was watching a [BazelCon 2024 video](https://www.youtube.com/watch?v=eiHTMU6a7uQ) when I noticed a couple of things:
- platform based flags are cool
- custom rules for platform transitions are not cool
- Bazel 8 was released yesterday
- if `platform_data` actually did what I wanted, we could use it...

This change migrates the repo to Bazel 8. A custom rule is still needed to set the linker script flag from within the same platform, but it's greatly simplified now. Using the new [symbolic macros](https://bazel.build/extending/macros) was a dead end and should wait for better support.

Remarkably this tested great on hardware. Appears to have made no change (truly, the best outcome).

### CODE REVIEW

**Comment by RyanDraves** [comment]
image

One day navigating Bazel docs or figuring out wtf to do will be easy. Today is not that day.

File: bzl/macros/pico.bzl [hunk 1]
```diff
@@ -28,13 +26,15 @@ def _pico_elf_and_bin(name, binary, linker_script, **kwargs):
...
+    platform_transition(
         name = name + ".bin",
-        binary = name + "_bin_intermediate",
+        dep = name + "_bin_intermediate",
+        platform = platform,
         **kwargs
     )
 
-def pico_project(name, srcs, deps, linker_script = "//emb/project/bootloader:application_linker_script", **kwargs):
+def pico_project(name, srcs, deps, platform = "//bzl/platforms:rp2040", **kwargs):
```
Comment (RyanDraves) [inline bzl/macros/pico.bzl:37]: I'd really love a way to say "of course this is `rp2040` platform, but you know the linker script flag that's set? Let's override it" but nothing I've searched for does that. In fact, if I kept the `pico.bzl` rule around, the docs specifically say that the platform flag will _override_ any transitions.

**Comment by RyanDraves** [comment]
Symbolic macros were a dead end and waste of time. I wasn't able to get one macro ported without some restriction, including the classic "all of your dependencies (and their dependencies) need to migrate first". :roll_eyes:

File: tools/bazelisms.md [hunk 2]
```diff
@@ -23,3 +23,14 @@ filegroup(
...
+## Platforms transitions
+
+Multi-platform builds require a platform transition on the targets to be built for other platforms. There are a few competing rules to be the "generic rule that implements a platform transition.
+
+The current state of affairs is:
+- [rules_platform](https://registry.bazel.build/modules/rules_platform)'s `platform_data` (binary targets only)
+- [aspect_bazel_lib](https://github.com/bazel-contrib/bazel-lib/blob/main/docs/transitions.md)'s `transitions.bzl` (binary targets, file targets, test targets)
+- `//bzl/rules:platform_transition.bzl` (file targets, CC libraries)
+
+It's a complete mess. If you want to customize flags within a platform, instead of binding all of them to the platform with platform flags, then custom rules for that are needed too.
```
Comment (michael-christen) [inline tools/bazelisms.md:36]: :scream:

**Pull request merged**